### PR TITLE
Use unified spending keys for spends & shielding.

### LINKED ISSUE #616: Enforce correspondence between sending and receiving account in `shield_transparent_funds`
Hmm; is it intended that `shield_transparent_funds` can shield funds from an _arbitrary_ account (specifically whichever account `sk` corresponds to) to `account`? Or should it only be shielding funds within `account` itself?

_Originally posted by @str4d in https://github.com/zcash/librustzcash/pull/615#discussion_r956467429_

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #616

### CODE REVIEW

File: zcash_client_backend/src/data_api.rs [hunk 1]
```diff
@@ -461,6 +467,14 @@ pub mod testing {
             Ok(HashMap::new())
         }
 
+        /// Returns all unified full viewing keys known to this wallet.
```
Comment (daira) [inline zcash_client_backend/src/data_api.rs:470]:
```suggestion
        /// Returns the account ID for a given unified full viewing key.
```

File: zcash_client_backend/src/data_api.rs [hunk 2]
```diff
@@ -461,6 +467,14 @@ pub mod testing {
...
+        fn get_account_for_ufvk(
+            &self,
+            _ufvk: &UnifiedFullViewingKey,
+        ) -> Result<Option<AccountId>, Self::Error> {
+            Ok(None)
```
Comment (daira) [inline zcash_client_backend/src/data_api.rs:475]: Should this be omitted until it is implemented?

File: zcash_client_backend/src/data_api.rs [hunk 3]
```diff
@@ -130,6 +130,12 @@ pub trait WalletRead {
         &self,
     ) -> Result<HashMap<AccountId, UnifiedFullViewingKey>, Self::Error>;
 
+    /// Returns all unified full viewing keys known to this wallet.
```
Comment (daira) [inline zcash_client_backend/src/data_api.rs:133]:
```suggestion
    /// Returns the account ID for a given unified full viewing key.
```

File: zcash_client_backend/src/data_api/wallet.rs [hunk 4]
```diff
@@ -127,7 +122,7 @@ where
 /// };
 /// use zcash_proofs::prover::LocalTxProver;
 /// use zcash_client_backend::{
-///     keys::sapling,
+///     keys::{UnifiedSpendingKey},
```
Comment (daira) [inline zcash_client_backend/src/data_api/wallet.rs:125]:
```suggestion
///     keys::UnifiedSpendingKey,
```

File: zcash_client_sqlite/src/wallet.rs [hunk 5]
```diff
@@ -352,6 +355,25 @@ pub(crate) fn get_unified_full_viewing_keys<P: consensus::Parameters>(
     Ok(res)
 }
 
+/// Returns the account id corresponding to a given [`UnifiedFullViewingKey`],
+/// if any.
```
Comment (daira) [inline zcash_client_sqlite/src/wallet.rs:359]: Document the circumstances under which there would be no account id for a UFVK.

File: zcash_primitives/CHANGELOG.md [hunk 6]
```diff
@@ -22,10 +21,14 @@ and this library adheres to Rust's notion of
...
 - Added in `zcash_primitives::zip32`
   - `ChainCode::as_bytes`
-  - `DiversifierKey::{from_bytes, as_bytes}`
   - `DiversifierIndex::{as_bytes}`
-  - `ExtendedSpendingKey::{from_bytes, to_bytes}`
   - Implementations of `From<u32>` and `From<u64>` for `DiversifierIndex`
+- `zcash_primitives::zip32::sapling` has been added and now contains 
+  all of the Sapling zip32 key types that were previously located in
+  `zcash_primitives::zip32` directly. The base `zip32` module reexports 
+  the moved types for backwards compability.
```
Comment (daira) [inline zcash_primitives/CHANGELOG.md:29]:
```suggestion
  the moved types for backwards compatibility.
```

File: zcash_primitives/src/zip32/sapling.rs [hunk 7]
```diff
@@ -0,0 +1,1811 @@
+//! Sapling key derivation according to ZIP 32 and ZIP 316
```
Comment (daira) [inline zcash_primitives/src/zip32/sapling.rs:1]: I have not checked that this is move-only.

**Review by daira — APPROVED** [review]
utACK with comments.

File: zcash_client_sqlite/src/wallet.rs (hunk 5 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet.rs:359]: The caller could conceivably call this with any arbitrary UFVK; does that need specific documentation?

File: zcash_client_backend/src/data_api.rs (hunk 1 above)
Comment (daira) [inline zcash_client_backend/src/data_api.rs:470]:
```suggestion
```

**Review by daira — COMMENTED** [review]
Note one additional inline comment (incorrect documentation in `MockWalletDb::get_account_for_ufvk`).

**Review by str4d — APPROVED** [review]
utACK b6e263d259dd0488a164e7691f5cc6e28f2aea58

File: zcash_client_backend/CHANGELOG.md [hunk 8]
```diff
@@ -124,6 +125,8 @@ and this library adheres to Rust's notion of
   - `decode_extended_spending_key`
   - `decode_extended_full_viewing_key`
   - `decode_payment_address`
+- `data_api::wallet::create_spend_to_address` has been modified to use a unified
+  spending key rather than a Sapling extended spending key.
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/CHANGELOG.md:129]: I will add that in a follow-up PR.

**Pull request merged**